rclcpp/include/rclcpp/memory_strategy.hpp

```diff
@@ -38,137 +36,171 @@ namespace memory_strategy
+  get_next_subscription(executor::AnyExecutable::SharedPtr any_exec,
+    const WeakNodeVector & weak_nodes) = 0;
+
+  virtual void
+  get_next_service(executor::AnyExecutable::SharedPtr any_exec,
+    const WeakNodeVector & weak_nodes) = 0;
+
+  virtual void
+  get_next_client(executor::AnyExecutable::SharedPtr any_exec,
+    const WeakNodeVector & weak_nodes) = 0;
+
+  static rclcpp::subscription::SubscriptionBase::SharedPtr
+  get_subscription_by_handle(void * subscriber_handle,
+    const WeakNodeVector & weak_nodes)
   {
-    switch (type) {
-      case HandleType::subscription_handle:
-        if (subscription_handles.size() < number_of_handles) {
-          subscription_handles.resize(number_of_handles, 0);
-        }
-        return static_cast<void **>(subscription_handles.data());
-      case HandleType::service_handle:
-        if (service_handles.size() < number_of_handles) {
-          service_handles.resize(number_of_handles, 0);
-        }
-        return static_cast<void **>(service_handles.data());
-      case HandleType::client_handle:
-        if (client_handles.size() < number_of_handles) {
-          client_handles.resize(number_of_handles, 0);
+    for (auto & weak_node : weak_nodes) {
+      auto node = weak_node.lock();
+      if (!node) {
+        continue;
+      }
+      for (auto & weak_group : node->get_callback_groups()) {
+        auto group = weak_group.lock();
+        if (!group) {
+          continue;
         }
-        return static_cast<void **>(client_handles.data());
-      case HandleType::guard_condition_handle:
-        if (number_of_handles > 2) {
-          throw std::runtime_error("Too many guard condition handles requested!");
+        for (auto & weak_subscription : group->get_subscription_ptrs()) {
+          auto subscription = weak_subscription.lock();
+          if (subscription) {
+            if (subscription->get_subscription_handle()->data == subscriber_handle) {
+              return subscription;
+            }
+            if (subscription->get_intra_process_subscription_handle() &&
+              subscription->get_intra_process_subscription_handle()->data == subscriber_handle)
+            {
+              return subscription;
+            }
+          }
         }
-        return guard_cond_handles.data();
-      default:
-        throw std::runtime_error("Unknown HandleType " + std::to_string(static_cast<int>(type)) +
-                ", could not borrow handle memory.");
+      }
     }
+    return nullptr;
   }
 
-  /// Return the memory borrowed in borrow_handles.
-  /**
-   * return_handles should always mirror the way memory was borrowed in borrow_handles.
-   * \param[in] The type of entity that this function is returning.
-   * \param[in] Pointer to the handles returned.
-   */
-  virtual void return_handles(HandleType type, void ** handles)
+  static rclcpp::service::ServiceBase::SharedPtr
+  get_service_by_handle(void * service_handle,
+    const WeakNodeVector & weak_nodes)
```

Comment:
Looks like all of the parameters could be on the same line without going over 100 characters.